basic_pipeline/04_StreamFormat.md

```diff
@@ -0,0 +1,136 @@
+Let's say that we are connecting two elements that process the video multimedia.
+The link is made between the pads which are working on raw video data.
+Here is where stream formats come up - they can be defined with the following constraints:
+
+- data format - in our case, we are having a raw video format
+- some additional constraints - i.e. [frame](../glossary/glossary.md#frame) resolution (480p) , framerate (30 fps) etc.
+
+The `:stream_format` action helps us find out if the given elements are capable to communicate with each other. Not only can we not send the data between the pads if the format they are expecting is different - we need to take into consideration some other constraints! We can think of a situation in which the _data format_ would be the same (i.e. raw video data), but the element which receives the data performs a much more complex computation on that data than the sender, and therefore cannot digest such a great amount of data as the sender is capable of transmitting. Then their stream formats wouldn't be compatible, which could be expressed by adding some constraint, i.e. framerate.
+
+Stream formats help us define a contract between elements and prevent us from connecting incompatible elements. That is why it is always better to define precise constraints rather than using `stream_format: :any`.
+
+## When are stream formats compatible?
+
+A comparison between formats is made when an input pad receives the `stream_format` action, and checks whether it matches its [accepted format](https://hexdocs.pm/membrane_core/Membrane.Pad.html#t:accepted_format/0).
+Here is how you define a stream format specification:
+
+1. First you need to specify the format module
+
+```elixir
+defmodule Formats.Raw do
+  defstruct [:pixel_format, :framerate, :width, :height]
+end
+```
+
+Module name defines the type of the format, however it is possible to pass some other options in a form of a struct. That is why we have defined a structure with the use of `defstruct`. Our format will be described with the following options:
+
+- :pixel_format - pixel format, i.e. [I420](https://en.wikipedia.org/wiki/Chroma_subsampling) ([YUV](https://en.wikipedia.org/wiki/YUV)) or RGB888
+- :framerate - number of frames per second, i.e. 30 (FPS)
+- :width - width of the picture in pixels, i.e. 480 (px)
+- :height - height of the picture in pixels, i.e. 300 (px)
+
+2. We specify the pad of the element with the format we have just defined, using the `:accepted_format` option. For the purpose of an example, let it be the `:input` pad:
+
+```elixir
+def_input_pad(:input,
+  demand_unit: :buffers,
+  accepted_format: [
+     {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 480, height: 300},
+     {Format.Raw, pixel_format: one_of([:I420, :I422]), framerate: range(30, 60), width: 720, height: 480}
+  ]
+)
```

Comment:
```suggestion
def_input_pad :input,
  demand_unit: :buffers,
  accepted_format: any_of(
  [
     %Format.Raw{pixel_format: pixel_format, framerate: framerate, width: 480, height: 300} when pixel_format in [:I420, :I422] and framerate >= 30 and framerate <= 60,
     %Format.Raw{pixel_format: pixel_format, framerate: range(30, 60), width: 720, height: 480} when pixel_format in [:I420, :I422] and framerate >= 30 and framerate <= 60 
  ]
```
Unfortunately, the caps mechanism has changed and now we use regular Elixir's matching mechanism instead of this custom parser with `one_of`, `range`, and similar functions. We need to reflect it in the description below (which will become much simpler since we can relate to a well-known mechanism :D As well as update the accepted format definition for our element)
You can read about it here: https://github.com/membraneframework/membrane_core/blob/168f57e12bf3280cda82b1116796151d4b1cb486/guides/upgrading/v0.11.md?plain=1#L224